src/scripts/main.js

```diff
@@ -1,3 +1,261 @@
+        return false;
+      }
+    }
+  }
+
+  for (let c = 0; c < columns; c++) {
+    const row = [board[0][c], board[1][c], board[2][c], board[3][c]];
+
+    for (let r = 0; r < row.length; r++) {
+      if (row[r] === row[r + 1]) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+function showLoseMessage() {
+  messageLose.classList.remove('hidden');
+}
+
+function updateCell(cell, num) {
+  cell.innerText = '';
+  cell.classList.value = '';
+  cell.classList.add('field-cell');
+
+  if (num > 0) {
+    cell.innerText = num;
+
+    cell.classList.add(`field-cell--${num}`);
+  }
+}
+
+document.addEventListener('keydown', (e) => {
+  switch (e.code) {
+    case 'ArrowLeft':
+      slideLeft();
+      break;
+    case 'ArrowRight':
+      slideRight();
+      break;
+    case 'ArrowUp':
+      slideUp();
+      break;
+    case 'ArrowDown':
+      slideDown();
+      break;
+  }
+
+  generateRandomCell();
+  scoreCell.innerText = score;
+  checkWin();
+
+  if (checkLose()) {
+    showLoseMessage();
+  }
+});
+
+function filterZero(row) {
+  return row.filter(num => num !== 0);
+}
+
+function slide(row) {
+  let currentRow = filterZero(row);
+
+  for (let i = 0; i < currentRow.length - 1; i++) {
+    if (currentRow[i] === currentRow[i + 1]) {
+      currentRow[i] *= 2;
+      currentRow[i + 1] = 0;
+      score += currentRow[i];
+    }
+  }
+
+  currentRow = filterZero(currentRow);
+
+  while (currentRow.length < columns) {
+    currentRow.push(0);
+  }
+
+  return currentRow;
+}
+
+function slideLeft() {
+  for (let r = 0; r < rows; r++) {
+    let row = board[r];
+
+    row = slide(row);
+    board[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      const cell = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateCell(cell, num);
+    }
+  }
+}
+
+function slideRight() {
+  for (let r = 0; r < rows; r++) {
+    let row = board[r];
+
+    row.reverse();
+
+    row = slide(row);
+
+    row.reverse();
+    board[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      const cell = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateCell(cell, num);
+    }
+  }
+}
+
+function slideUp() {
+  for (let c = 0; c < columns; c++) {
+    let row = [board[0][c], board[1][c], board[2][c], board[3][c]];
+
+    row = slide(row);
+
+    for (let r = 0; r < rows; r++) {
+      board[r][c] = row[r];
+
+      const cell = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateCell(cell, num);
+    }
+  }
+}
+
+function slideDown() {
+  for (let c = 0; c < columns; c++) {
+    let row = [board[0][c], board[1][c], board[2][c], board[3][c]];
+
+    row.reverse();
+    row = slide(row);
+    row.reverse();
+
+    for (let r = 0; r < rows; r++) {
+      board[r][c] = row[r];
+
+      const cell = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateCell(cell, num);
+    }
```

Comment:
this code is duplicated in 4 functions, so make sense to move it to the separate function 
**for (let r = 0; r < rows; r++) {
      board[r][c] = row[r];

      const cell = document.getElementById(`${r}-${c}`);
      const num = board[r][c];

      updateCell(cell, num);
    }**
    please check other code duplications